Make privacy visitor use types more (instead of HIR)

oli-obk opened the pull request:
r? @petrochenkov 

This is a prerequisite to normalizing projections, as otherwise we have too many invalid bound vars (hir_ty_to_ty is creating types that have bound vars, but no binder).

The commits are still chaotic, I'm gonna clean them up, but I just wanted to let you know about the general direction and wondering if we could land this before adding normalization, as normalization is where behavioral changes happen, and I'd like to keep that part as minimal as possible.

[context can be found on zulip](https://rust-lang.zulipchat.com/#narrow/stream/315482-t-compiler.2Fetc.2Fopaque-types/topic/weak.20type.20aliases.20and.20privacy)

petrochenkov commented:
cc @Bryanskiy

petrochenkov commented on compiler/rustc_privacy/src/lib.rs:
```diff
@@ -1347,16 +1389,34 @@ impl<'tcx> Visitor<'tcx> for TypePrivacyVisitor<'tcx> {
         let orig_current_item = mem::replace(&mut self.current_item, item.owner_id.def_id);
         let old_maybe_typeck_results = self.maybe_typeck_results.take();
-        intravisit::walk_item(self, item);
+        rustc_ty_utils::sig_types::walk_types(self.tcx(), item.owner_id.def_id, self);
```
I also don't understand how `walk_types` checks types in signatures, and why the additional code below is necessary.

If we have a, for example, function
```
// Assume `Alias` always results in `u8`, regardless of type arguments
fn foo(arg: Alias<Priv, Pub>) {}
```
we need to recursively walk HIR types in its signature, normalize them, and check for privacy, i.e.

- `Alias<Priv, Pub>` -> normalize -> `u8` -> public, ok
- `Priv` -> normalize -> `Priv` -> private, error
- `Pub` -> normalize -> `Pub` -> public, ok

(Setting good spans important, but it wasn't the primary point of visiting HIR here.)

`walk_types` seems to only take the top level type of `arg`, and then use `TypeVisitable` on it (i.e. on `u8`) and ignore `Priv` and `Pub`.
I added tests for cases like this back in 2017, not sure why they don't fail.

oli-obk replied:
Walk_types does not normalize. When we start normalizing, we'll run on both the normalized type and the unnormalized type to catch everything as you described

oli-obk replied:
> Now they are only check in signatures of free items, as I understand.

Yea I was wondering about that. I'll add some tests

oli-obk replied:
We have all the tests we could want. There's also `PrivateItemsInPublicInterfacesChecker` which does something similar, but apparently not exactly the same thing? I'm confused 😆

oli-obk replied:
It's not clear to me what `check_mod_privacy` and `check_private_in_public` are supposed to do differently and why they do a lot of the same things.

petrochenkov replied:
Briefly, `check_mod_privacy` checks uses, it's precise (report errors) and pretty conservative / future-proofing (e.g. it's not entirely clear whether we *should* report errors for `Priv` in https://github.com/rust-lang/rust/pull/113671#discussion_r1263847616).

`check_private_in_public` checks definitions trying to predict which definitions will later result in `check_mod_privacy` errors on use, it's imprecise (reports lints) and misses some cases (allow leaking private stuff).

There's also some legacy stuff in there.
Everything will become a bit simpler when https://github.com/rust-lang/rust/pull/113126 lands, if you want we can delay this PR until then.

Further simplifications / deduplication may also be unlocked by correctly normalizing all associated types.

petrochenkov commented:
(I also don't understand all the new binder play here, but I assume it's not something complex, and you know what you are doing.)

petrochenkov commented:
>wondering if we could land this before adding normalization, as normalization is where behavioral changes happen, and I'd like to keep that part as minimal as possible.

Yes, that's quite reasonable.

oli-obk commented on compiler/rustc_privacy/src/lib.rs:
```diff
@@ -176,6 +192,20 @@ where
+    }
+
+    fn visit_predicate(&mut self, p: ty::Predicate<'tcx>) -> ControlFlow<Self::BreakTy> {
+        self.visit_clause(p.as_clause().unwrap())
```
The visit_clause has special logic for traits. This type visitor would not use it and thus only look at the types in the trait ref arguments, instead of discovering the trait ref's trait and marking it('s def id) as seen

petrochenkov replied:
How did this work previously?
Does this change fix some pre-existing issue?

I tried to remove this change and got some missing errors.
But without this PR, on master, those errors are reported even without the `visit_predicate` override.

petrochenkov replied:
FWIW, I still cannot find where the *real* default implementation of `fn visit_predicate` lives, there are too many abstraction layers.

oli-obk commented on compiler/rustc_privacy/src/lib.rs:
```diff
@@ -356,10 +380,11 @@ trait VisibilityLike: Sized {
         effective_visibilities: &EffectiveVisibilities,
     ) -> Self {
-        let mut find = FindMin::<_, SHALLOW> { tcx, effective_visibilities, min: Self::MAX };
+        let infcx = tcx.infer_ctxt().build();
```
We could use a single inferctxt always. Or generate it whenever we create the universe table. I'll convert to doing the latter. This inferctxt logic is leftover from a previous version of this PR

oli-obk replied:
I just removed it for now, we only need it when we actually normalize

oli-obk commented:
> (I also don't understand all the new binder play here, but I assume it's not something complex, and you know what you are doing.)

I'll pull it into a separate commit. It's not necessary until we do normalization, and it's indeed just a mechanical refactoring to not lose the binders

bors commented:
:umbrella: The latest upstream changes (presumably #113591) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

oli-obk commented on compiler/rustc_privacy/src/lib.rs:
```diff
@@ -232,11 +247,15 @@ where
+                    ty::Weak => (data.def_id, "type alias"),
+                    // Trait associated types don't have visibility, they are
+                    // visible if their parent trait is.
+                    ty::Projection => (tcx.parent(data.def_id), "trait"),
```
d9726ca76df26718ce40cfb4d2eaf9b7076e0226 shows why we need it

oli-obk commented on compiler/rustc_privacy/src/lib.rs:
```diff
@@ -207,22 +237,7 @@ where
-                    return ControlFlow::Continue(());
-                }
-                // This will also visit substs if necessary, so we don't need to recurse.
-                return self.visit_projection_ty(proj);
```
yes

oli-obk committed:
Stop using `hir_ty_to_ty` in rustc_privacy

bors commented:
:umbrella: The latest upstream changes (presumably #113522) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

petrochenkov commented:
https://github.com/rust-lang/rust/pull/113126 has landed.
@rustbot author

oli-obk commented:
@rustbot ready

not sure about the last commit, but I added it to show why I'm special casing projections

petrochenkov commented on compiler/rustc_privacy/src/lib.rs:
```diff
@@ -232,11 +221,21 @@ where
+                    ty::Weak => Some("type alias"),
+                    // Trait associated types don't have visibility, they are
+                    // visible if their parent trait is.
+                    ty::Projection => None,
```
```suggestion
                    ty::Projection => Some("associated type"),
```
Not sure what the last commit intends to demonstrate.
I didn't suggest skipping this logic for associated types, I only suggested to use their own def ids.

Associated type def ids do have (semantic) visibility (the comment above is wrong in this regard), but that visibility is indeed inherited from the trait (the comment above is right in that regard).

oli-obk replied:
yea that makes sense. I added a change that just exposes associated types directly here. I will edit git history at the end before merging the PR

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6143523591/job/16667142542) [(plain)](https://github.com/rust-lang/rust/commit/3e1da2298e7da2d83fa364fac51f005d208f968b/checks/16667142542/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_414945f9-2bb6-46dc-bcf8-53017844b7c4
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=normalize_weak_tys
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_414945f9-2bb6-46dc-bcf8-53017844b7c4
GITHUB_REF=refs/pull/113671/merge
GITHUB_REF_NAME=113671/merge
GITHUB_REF_PROTECTED=false
---
failures:

---- [ui] tests/ui/privacy/associated-item-privacy-trait.rs stdout ----

error: /checkout/tests/ui/privacy/associated-item-privacy-trait.rs:25: unexpected error: '25:34: 25:58: associated type `PrivTr::AssocTy` is private'
error: 1 unexpected errors found, 0 expected errors not found
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/privacy/associated-item-privacy-trait.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/privacy/associated-item-privacy-trait" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/privacy/associated-item-privacy-trait/auxiliary"
unexpected errors (from JSON output): [
unexpected errors (from JSON output): [
    Error {
        line_num: 25,
        kind: Some(
            Error,
        ),
        msg: "25:34: 25:58: associated type `PrivTr::AssocTy` is private",
]

thread '[ui] tests/ui/privacy/associated-item-privacy-trait.rs' panicked at src/tools/compiletest/src/runtest.rs:1809:13:
explicit panic
explicit panic

---- [ui] tests/ui/privacy/private-in-public.rs stdout ----

error: /checkout/tests/ui/privacy/private-in-public.rs:106: unexpected warning: '106:5: 106:44: associated type `aliases_pub::PrivTr::Assoc` is more private than the item `aliases_pub::f3` [private_interfaces]'

error: /checkout/tests/ui/privacy/private-in-public.rs:134: unexpected warning: '134:5: 134:44: associated type `aliases_priv::PrivTr::Assoc` is more private than the item `aliases_priv::f3` [private_interfaces]'
error: 2 unexpected errors found, 0 expected errors not found
status: exit status: 0
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/privacy/private-in-public.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/privacy/private-in-public" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/privacy/private-in-public/auxiliary"
unexpected errors (from JSON output): [
unexpected errors (from JSON output): [
    Error {
        line_num: 106,
        kind: Some(
            Warning,
        ),
        msg: "106:5: 106:44: associated type `aliases_pub::PrivTr::Assoc` is more private than the item `aliases_pub::f3` [private_interfaces]",
    Error {
        line_num: 134,
        kind: Some(
            Warning,
            Warning,
        ),
        msg: "134:5: 134:44: associated type `aliases_priv::PrivTr::Assoc` is more private than the item `aliases_priv::f3` [private_interfaces]",
]

thread '[ui] tests/ui/privacy/private-in-public.rs' panicked at src/tools/compiletest/src/runtest.rs:1809:13:
explicit panic
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #116163) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

oli-obk commented:
This is ready for review again

bors commented:
:umbrella: The latest upstream changes (presumably #116455) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

wesleywiser commented:
> I still have other questions, but it would be good to sort out the last commit first.

@petrochenkov it's not clear to me, did those questions get resolved or are there still open threads of discussion? 

Thanks!

petrochenkov commented:
I was just busy for the most of September and accumulated a large review queue.
This PR is the last in the queue and I'm actually reviewing it today.

petrochenkov commented on tests/ui/privacy/private-in-public.rs:
```diff
@@ -104,8 +104,8 @@ mod aliases_pub {
     pub fn f3(arg: <Priv as PrivTr>::Assoc) {}
-    //~^ WARNING trait `aliases_pub::PrivTr` is more private than the item `aliases_pub::f3`
-    //~| WARNING type `aliases_pub::Priv` is more private than the item `aliases_pub::f3`
+    //~^ WARNING type `aliases_pub::Priv` is more private than the item `aliases_pub::f3`
```
Nit: unnecessary annotation order change here and below.

petrochenkov commented on compiler/rustc_privacy/src/lib.rs:
```diff
@@ -1193,7 +1196,7 @@ impl<'tcx> Visitor<'tcx> for TypePrivacyVisitor<'tcx> {
             // FIXME: This is very ineffective. Ideally each HIR type should be converted
             // into a semantic type only once and the result should be cached somehow.
-            if self.visit(rustc_hir_analysis::hir_ty_to_ty(self.tcx, hir_ty)).is_break() {
+            if self.visit(rustc_hir_analysis::hir_ty_to_ty(self.tcx(), hir_ty)).is_break() {
```
```suggestion
            if self.visit(rustc_hir_analysis::hir_ty_to_ty(self.tcx, hir_ty)).is_break() {
```

petrochenkov commented on compiler/rustc_privacy/src/lib.rs:
```diff
@@ -1358,7 +1361,20 @@ impl<'tcx> Visitor<'tcx> for TypePrivacyVisitor<'tcx> {
+                self.visit_def_id(trait_ref.def_id, "trait", &trait_ref.print_only_trait_path());
+            }
+            _ => {}
+        }
```
Wait, if `fn visit_ty` with `hir_ty_to_ty` still exists, then why this change is needed?
All the signature types will be visited by `visit_ty` anyway.

oli-obk replied:
Oh yea that was all just dead code. I removed it

petrochenkov commented:
Can we merge everything besides the `fn visit_item` change in https://github.com/rust-lang/rust/pull/113671/commits/f507b14b3dbc7b4e164ed8a2a900a051ca1c79d6 first?
I'm still not sure that it's doing something right, but I'll return to it later.
@rustbot author

oli-obk commented on compiler/rustc_ty_utils/src/sig_types.rs:
```diff
@@ -0,0 +1,129 @@
+use rustc_span::Span;
+use rustc_type_ir::visit::TypeVisitable;
+
+pub trait Spanned<'tcx> {
```
addressed in https://github.com/rust-lang/rust/pull/117076

oli-obk commented on compiler/rustc_ty_utils/src/opaque_types.rs:
```diff
@@ -273,25 +276,14 @@ fn opaque_types_defined_by<'tcx>(tcx: TyCtxt<'tcx>, item: LocalDefId) -> &'tcx [
+        | DefKind::Static(_)
+        | DefKind::Const
+        | DefKind::AssocConst
+        | DefKind::AnonConst => {
```
addressed in https://github.com/rust-lang/rust/pull/117076

petrochenkov commented:
Blocked on https://github.com/rust-lang/rust/pull/117076.
@rustbot blocked

bors commented:
:umbrella: The latest upstream changes (presumably #117405) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

petrochenkov commented:
(I'll return to this around early December, I think.)

oli-obk commented:
I'm on leave until January anyway 😃

petrochenkov commented:
Ok, what I see here now:
- Types in signatures are not checked in trait/impl/foreign items and in fields.
- Types are not checked before type alias normalization, i.e. type `IgnoreArguments<PrivTy>` won't be reported, but it was reported previously. (Or at least I fail to see how they could be reported.)
@rustbot author

bors commented:
:umbrella: The latest upstream changes (presumably #120309) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

oli-obk commented:
> Types in signatures are not checked in trait/impl/foreign items and in fields.

They are now. Nothing changed, so they must have been checked before, probably by going down the HIR route in the recursive visitor. Now they are taking the ty route.

> Types are not checked before type alias normalization, i.e. type `IgnoreArguments<PrivTy>` won't be reported, but it was reported previously. (Or at least I fail to see how they could be reported.)

that's because those type aliases are prevented with hard errors: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=e143c6075b63bc855e751c7053bf2329

petrochenkov commented:
>that's because those type aliases are prevented with hard errors: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=e143c6075b63bc855e751c7053bf2329

Interesting.
That's true for free type aliases, but not for associated type aliases, e.g. in this example
```rust
#![feature(type_alias_impl_trait)]
#![allow(unused)]

mod m {
    struct Priv;
    pub type Leak = Priv;
}

trait Trait {
    type A<T>;
}

impl Trait for u8 {
    type A<T> = u8;
}

fn check() -> <u8 as Trait>::A<m::Leak> {
    0
}
```

This example still produces an error on this PR's branch though, not sure why.

petrochenkov commented on compiler/rustc_privacy/src/lib.rs:
```diff
@@ -1085,71 +1098,39 @@ impl<'tcx> Visitor<'tcx> for TypePrivacyVisitor<'tcx> {
     }
 
     fn visit_trait_ref(&mut self, trait_ref: &'tcx hir::TraitRef<'tcx>) {
         self.span = trait_ref.path.span;
```
It's probably unnecessary to override `visit_trait_ref` now, because this assignment is dead - the span is always overwritten by a more precise span in `walk_trait_ref`.

petrochenkov commented on compiler/rustc_privacy/src/lib.rs:
```diff
@@ -13,6 +13,7 @@ extern crate tracing;
 
 mod errors;
 
+use hir::ItemKind;
```
Nit: non-standard imports was auto-inserted by an editor.

petrochenkov commented:
>They are now. Nothing changed, so they must have been checked before, probably by going down the HIR route in the recursive visitor. Now they are taking the ty route.

Fields are still not visited.
This snippet doesn't error on this PR.
```rust
mod m {
    struct Priv;
    pub type Leak = Priv;
}

struct S {
    field: m::Leak
}
```

I don't like how easy it is to miss something with this approach :(
Previously `visit_ty` covered everything automatically.

petrochenkov commented:
>I don't like how easy it is to miss something with this approach :(
>Previously `visit_ty` covered everything automatically.

Is it possible for `walk_types` to e.g. collect a set of visited `HirId`s, and then for `visit_ty`/`visit_trait_ref` to panic on anything that is not in that set?

oli-obk commented:
> I don't like how easy it is to miss something with this approach :(
> Previously `visit_ty` covered everything automatically.

actually that is because of a change in the `walk_types` logic where we used to walk into fields, but now stopped doing that *recursively* but I also accidentally made it stop doing that on the outermost level (struct/union/enum declarations).

I don't think this change is more fragile, it's just the conversion from one scheme to another that is bug-prone.

oli-obk committed:
Associated types in traits don't necessarily have a type that we can visit

This prevents ICEs from happening in the future when this code path is actually used

oli-obk committed:
extern types don't have any types to visit

oli-obk committed:
Actually walk fields of Adt definitions

oli-obk commented:
> This example still produces an error on this PR's branch though, not sure why.

we walk the arguments of the projection. We don't normalize, so if there's a private type behind the projection then we'll miss it. Not sure that can happen, as the impl would probably be the problematic one then.

oli-obk committed:
Remove now-useless method override

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7816813102/job/21323274318) [(plain)](https://github.com/rust-lang/rust/commit/5c25de67d49014397d5c09bcac5f2770b9fda1be/checks/21323274318/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_b1bbe769-560a-4ac8-8eb4-0901276e05ee
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=normalize_weak_tys
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_b1bbe769-560a-4ac8-8eb4-0901276e05ee
GITHUB_REF=refs/pull/113671/merge
GITHUB_REF_NAME=113671/merge
GITHUB_REF_PROTECTED=false
---
#12 writing image sha256:854ec7adf5d4d0780139e54e071c61f178eb1393b02ffdcda487d613a7c1e599 done
#12 naming to docker.io/library/rust-ci done
#12 DONE 10.0s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Wed Feb  7 15:05:38 UTC 2024
  network time: Wed, 07 Feb 2024 15:05:38 GMT
  network time: Wed, 07 Feb 2024 15:05:38 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'build.optimized-compiler-builtins', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
    |
71  | trait DefIdVisitor<'tcx> {
    |       ------------ method in this trait
...
101 |     fn visit_projection_ty(&mut self, projection: ty::AliasTy<'tcx>) -> ControlFlow<Self::BreakTy> {
    |
    = note: `-D dead-code` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(dead_code)]`

```

</details>

oli-obk committed:
Remove dead code

petrochenkov commented:
> > This example still produces an error on this PR's branch though, not sure why.
> 
> [we walk the arguments of the projection](https://github.com/rust-lang/rust/blob/5c25de67d49014397d5c09bcac5f2770b9fda1be/compiler/rustc_privacy/src/lib.rs#L141). We don't normalize, so if there's a private type behind the projection then we'll miss it. Not sure that can happen, as the impl would probably be the problematic one then.

From what I understand, the compiler will eventually keep all type aliases in the `ty` representation.
That will eliminate the risks related to `AliasIgnoringArguments<PrivTy>` in all cases, we'll likely just need to check privacy at every "normalization step", instead of "every HIR type" like before this PR.

petrochenkov commented:
Could you add the case from https://github.com/rust-lang/rust/pull/113671#issuecomment-1931903873 to tests (if there's no equivalent)?
r=me after that.
@rustbot author

oli-obk committed:
Add some tests for associated type normalization edge cases

oli-obk commented:
@bors r=petrochenkov

bors commented:
:pushpin: Commit d80d7ea1e325ded1b8b57204cbf98e65b6e0957c has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov d80d7ea1e325ded1b8b57204cbf98e65b6e0957c -->
<!-- homu: {"type":"Approved","sha":"d80d7ea1e325ded1b8b57204cbf98e65b6e0957c","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
